cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp

```diff
@@ -79,13 +80,41 @@ inline sycl::event compute_covariance(sycl::queue& q,
         cgh.depends_on(deps);
         cgh.parallel_for(range, [=](sycl::item<2> id) {
             const std::int64_t gi = id.get_linear_id();
-            const std::int64_t i = id.get_id(0);
-            const std::int64_t j = id.get_id(1);
 
-            if (i < p && j < p) {
-                cov_ptr[gi] -= inv_n * sums_ptr[i] * sums_ptr[j];
-                cov_ptr[gi] *= multiplier;
-            }
+            cov_ptr[gi] -= inv_n * sums_ptr[id.get_id(0)] * sums_ptr[id.get_id(1)];
+            cov_ptr[gi] *= multiplier;
+        });
+    });
+}
+
+template <typename Float>
+inline sycl::event compute_covariance_centered(sycl::queue& q,
+                                               std::int64_t row_count,
+                                               const ndview<Float, 1>& sums,
+                                               ndview<Float, 2>& cov,
+                                               bool bias,
+                                               const event_vector& deps) {
+    ONEDAL_ASSERT(sums.has_data());
+    ONEDAL_ASSERT(cov.has_mutable_data());
+    ONEDAL_ASSERT(cov.get_dimension(0) == cov.get_dimension(1), "Covariance matrix must be square");
+    ONEDAL_ASSERT(is_known_usm(q, sums.get_data()));
+    ONEDAL_ASSERT(is_known_usm(q, cov.get_mutable_data()));
+
+    const std::int64_t n = row_count;
+    const std::int64_t p = sums.get_count();
+    const Float inv_n = Float(1.0 / double(n));
+    const Float inv_n1 = (n > 1) ? Float(1.0 / double(n - 1)) : Float(1);
+    const Float multiplier = bias ? inv_n : inv_n1;
+
+    Float* cov_ptr = cov.get_mutable_data();
+
+    return q.submit([&](sycl::handler& cgh) {
+        const auto range = make_range_2d(p, p);
+
+        cgh.depends_on(deps);
+        cgh.parallel_for(range, [=](sycl::item<2> id) {
+            const std::int64_t gi = id.get_linear_id();
+            cov_ptr[gi] *= multiplier;
```

Comment:
Wouldn't it be better to have 1d range of size p * p instead?

Reply:
It makes sense! Thanks!